### PR TITLE
Update tests to compare at Model level #456

### LINKED ISSUE #456: Update tests to compare at Model level
The current test code base has the following code that tests for equality for the model's components: 
```
assertEquals(expectedAddressBook, model.getAddressBook());
assertEquals(expectedFilteredList, model.getFilteredPersonList());
```

A more compact way would be to `assertEquals(expectedModel, model);`

Let's update the existing test code base by implementing Model#equals to support the above mentioned implementation.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #456 

Proposed merge commit message:
```
Update tests to use assertEquals(Model, Model)

Tests check for equality of the individual components of Model 
(AddressBook and FilteredList). The following code snippet shows an 
example:

assertEquals(expectedAddressBook, model.getAddressBook());
assertEquals(expectedFilteredList, model.getFilteredPersonList());

This can be simplified by checking for equality of the entire Model
directly using ModelManager#equals().

Let's implement Model#equals() and update the tests to check for
equality of Model using assertEquals(expectedModel, model).
```

### CODE REVIEW

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 1]
```diff
@@ -178,12 +173,8 @@ private void assertParseException(String inputCommand, String expectedMessage) {
...
         }
 
-        //Confirm the ui display elements should contain the right data
-        assertEquals(expectedShownList, model.getFilteredPersonList());
-
-        //Confirm the state of data (saved and in-memory) is as expected
-        assertEquals(expectedAddressBook, model.getAddressBook());
-        assertEquals(expectedAddressBook, latestSavedAddressBook);
+        assertEquals(expectedModelManager, model);
+        assertEquals(expectedModelManager.getAddressBook(), latestSavedAddressBook);
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Why is this line needed? Shouldn't `assertEquals(expectedModelManager, model);` be already comparing the address book?

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 2]
```diff
@@ -149,25 +148,21 @@ private void assertParseException(String inputCommand, String expectedMessage) {
...
-     *      - {@code expectedAddressBook} was saved to the storage file. <br>
+     * and also confirms that the following two parts of the LogicManager object's state are as expected:<br>
+     *      - the internal model manager data are same as those in the {@code expectedModelManager} <br>
+     *      - {@code expectedModelManager}'s address book was saved to the storage file. <br>
      */
-    private <T> void assertCommandBehavior(Class<T> expectedException,
-            String inputCommand, String expectedMessage, ReadOnlyAddressBook expectedAddressBook,
-            List<? extends ReadOnlyPerson> expectedShownList) {
+    private <T> void assertCommandBehavior(Class<T> expectedException, String inputCommand,
+                                           String expectedMessage, Model expectedModelManager) {
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: `expectedModel`

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 3]
```diff
@@ -120,18 +121,16 @@ public void execute_invalid() {
     /**
      * Executes the command, confirms that no exceptions are thrown and that the result message is correct.
      * Also confirms that both the 'address book' and the 'last shown list' are as specified.
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Can update the header comments accordingly. Now we are comparing the entire model.

**Comment by yamgent** [comment]
Also you didn't submit a CanIHasReview. :(

File: src/test/java/seedu/address/logic/LogicManagerTest.java (hunk 1 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: because header comments

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/2] [Implement ModelManager\#equals](https://github.com/se-edu/addressbook-level4/pull/458/commits/d1dc430e948627f1be5d5598ce6039c011f100ef)
* [v1 2/2] [Update tests to use assertEquals\(Model\, Model\)](https://github.com/se-edu/addressbook-level4/pull/458/commits/569fcdffc2a14d1e3610e0fe2572cf7acc6b8147)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...569fcdffc2a14d1e3610e0fe2572cf7acc6b8147))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/458/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 4]
```diff
@@ -411,14 +386,11 @@ public void execute_find_onlyMatchesFullWordsInNames() throws Exception {
...
-        List<Person> expectedList = helper.generatePersonList(pTarget1, pTarget2);
+        Model expectedModel = new ModelManager(helper.generateAddressBook(fourPersons), new UserPrefs());
+        expectedModel.updateFilteredPersonList(new HashSet<>(Collections.singletonList("KEY")));
         helper.addToModel(model, fourPersons);
-
         assertCommandSuccess(FindCommand.COMMAND_WORD + " KEY",
-                Command.getMessageForPersonListShownSummary(expectedList.size()),
-                expectedAb,
-                expectedList);
+                Command.getMessageForPersonListShownSummary(2), expectedModel);
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Don't use magic numbers.

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 5]
```diff
@@ -430,14 +402,11 @@ public void execute_find_isNotCaseSensitive() throws Exception {
...
-        AddressBook expectedAb = helper.generateAddressBook(fourPersons);
-        List<Person> expectedList = fourPersons;
+        Model expectedModel = new ModelManager(helper.generateAddressBook(fourPersons), new UserPrefs());
         helper.addToModel(model, fourPersons);
 
         assertCommandSuccess(FindCommand.COMMAND_WORD + " KEY",
-                Command.getMessageForPersonListShownSummary(expectedList.size()),
-                expectedAb,
-                expectedList);
+                Command.getMessageForPersonListShownSummary(4), expectedModel);
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Same.

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 6]
```diff
@@ -449,14 +418,12 @@ public void execute_find_matchesIfAnyKeywordPresent() throws Exception {
...
-        List<Person> expectedList = helper.generatePersonList(pTarget1, pTarget2, pTarget3);
+        Model expectedModel = new ModelManager(helper.generateAddressBook(fourPersons), new UserPrefs());
+        expectedModel.updateFilteredPersonList(new HashSet<>(Arrays.asList("key", "rAnDoM")));
         helper.addToModel(model, fourPersons);
 
         assertCommandSuccess(FindCommand.COMMAND_WORD + " key rAnDoM",
-                Command.getMessageForPersonListShownSummary(expectedList.size()),
-                expectedAb,
-                expectedList);
+                Command.getMessageForPersonListShownSummary(3), expectedModel);
```
Comment (yamgent) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Same

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/2] [Implement ModelManager\#equals\(\)](https://github.com/se-edu/addressbook-level4/pull/458/commits/cd073a57085c6993ae05f9ab33dd68554c8fa32e)
* [v2 2/2] [Update tests to use assertEquals\(Model\, Model\)](https://github.com/se-edu/addressbook-level4/pull/458/commits/a1d8ba150a099d36a31443ba8a17e1d6ac6d8e9a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...a1d8ba150a099d36a31443ba8a17e1d6ac6d8e9a)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/236e767f188c90df5a425f19e82dd83da376cf8b~1...236e767f188c90df5a425f19e82dd83da376cf8b))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/458/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/model/ModelManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        // different types -> returns false
+        assertFalse(modelManager.equals(5));
+
+        // different addressBook -> returns false
+        assertFalse(modelManager.equals(new ModelManager(new AddressBook(), userPrefs)));
+
+        // different userPrefs -> returns true
+        UserPrefs differentUserPrefs = new UserPrefs();
+        differentUserPrefs.setAddressBookName("differentName");
+        assertTrue(modelManager.equals(new ModelManager(addressBook, differentUserPrefs)));
```
Comment (damithc) [inline src/test/java/seedu/address/model/ModelManagerTest.java]: Should we test for differences in `filteredList` too?

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 8]
```diff
@@ -119,55 +120,46 @@ public void execute_invalid() {
...
 
     /**
      * Executes the command, confirms that the result message is correct and that the expected exception is thrown,
-     * and also confirms that the following three parts of the LogicManager object's state are as expected:<br>
-     *      - the internal address book data are same as those in the {@code expectedAddressBook} <br>
-     *      - the backing list shown by UI matches the {@code shownList} <br>
-     *      - {@code expectedAddressBook} was saved to the storage file. <br>
+     * and also confirms that the following two parts of the LogicManager object's state are as expected:<br>
+     *      - the internal model manager data are same as those in the {@code expectedModel} <br>
+     *      - {@code expectedModel}'s address book was saved to the storage file. <br>
```
Comment (damithc) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: last `<br>` not needed?

File: src/test/java/seedu/address/model/ModelManagerTest.java (hunk 7 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/model/ModelManagerTest.java]: noted prof, I have added it, but there's a need to reset the `filteredList` afterwards (see line 46). I think the comment looks weird though.

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/2] [Implement ModelManager\#equals\(\)](https://github.com/se-edu/addressbook-level4/pull/458/commits/ec9802858a993112faf0810e9e242a658d3f5d4b)
* [v3 2/2] [Update tests to use assertEquals\(Model\, Model\)](https://github.com/se-edu/addressbook-level4/pull/458/commits/8d4cd683088b0983fc4d157bce8b539a4deff9ac)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...8d4cd683088b0983fc4d157bce8b539a4deff9ac)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/af7f8ef64dfefb83f31a22f0fac1e8aaf5356e80~1...af7f8ef64dfefb83f31a22f0fac1e8aaf5356e80))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/458/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/model/ModelManagerTest.java (hunk 7 above)
Comment (damithc) [inline src/test/java/seedu/address/model/ModelManagerTest.java]: May be it can be the last test inside that method?

Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/model/ModelManagerTest.java]: I'm thinking of that too, however it seems a little unsafe because someone may come adding another test inside the method, and it will take a while for the person to debug :<

**Comment by Zhiyuan-Amos** [comment]
@yamgent last call? :>

**Comment by yamgent** [comment]
@MightyCupcakes A reminder to review the PR if you have the time. :)
